tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+                                           int total_tiles_count,
+                                           const T* input,
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+
+    static_assert((TileLongSide & (TileLongSide - 1)) == 0,
+      "The length of the longer side of the tile is always a power of 2.");
+    bool request_satisfied = max(tile_size_i, tile_size_j) <= TileLongSide &&
+                             min(tile_size_i, tile_size_j) <= TileShortSide;
+
+    if (request_satisfied) {
+      const int NumThread = TileLongSide;
+      if (tile_size_i <= TileLongSide && tile_size_j <= TileShortSide) {
+        SwapDimension1And2InTensor3UsingTiles<
+            T, NumThread, TileLongSide,
+            TileShortSide><<<total_tiles_count, NumThread, 0, d.stream()>>>(
+            input, input_dims, output);
+      }
+      else if (tile_size_j <= TileLongSide && tile_size_i <= TileShortSide) {
+        SwapDimension1And2InTensor3UsingTiles<
+            T, NumThread, TileShortSide,
+            TileLongSide><<<total_tiles_count, NumThread, 0, d.stream()>>>(
+            input, input_dims, output);
+      }
+      return;
+    }
+
+    // Kernel is not launched, meaning the launch configuration is not
+    // satisfied.
+    const bool long_side_request_not_satisfied =
+        max(tile_size_i, tile_size_j) > TileLongSide;
+
+    // Increase launch parameters and try again.
+    if (long_side_request_not_satisfied) {
+      BatchNarrowMatrixTransposeDispatcher<T, TileLongSide * 2, TileShortSide>::
+          DoIt(d, tile_size_i, tile_size_j,
+                                       total_tiles_count, input, input_dims,
+                                       output);
+    } else {
+      BatchNarrowMatrixTransposeDispatcher<T, TileLongSide, TileShortSide + 1>::
+          DoIt(d, tile_size_i, tile_size_j,
+                                       total_tiles_count, input, input_dims,
+                                       output);
+    }
+  }
+};
+
+template <typename T, int TileLongSide, int TileShortSide>
+struct BatchNarrowMatrixTransposeDispatcher<
+    T, TileLongSide, TileShortSide,
+    typename std::enable_if<TileSizePossibilityFrontierCheck(
+                                TileLongSide, TileShortSide, sizeof(T), 0),
+                            void>::type> {
+  static void DoIt(const GPUDevice& d, int tile_size_i,
+                                           int tile_size_j,
+                                           int total_tiles_count,
+                                           const T* input,
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+
+    static_assert((TileLongSide & (TileLongSide - 1)) == 0,
+      "The length of the longer side of the tile is always a power of 2.");
+
+    const int NumThread = TileLongSide;
```

Comment:
Huh, I am surprised this works with regular `const` and not constexpr, since you use it as a template param.  Probably better to use constexpr to drive the point home, though.